103BinaryTreeZigzagLevelOrderTraversal.md

```diff
@@ -0,0 +1,287 @@
+以下のコードは全てGoのデフォルトフォーマッターgofmtにかけてあります。
+
+```Go
+// Definition for a binary tree node.
+type TreeNode struct {
+	Val   int
+	Left  *TreeNode
+	Right *TreeNode
+}
+```
+
+### Step 1
+- 方針: BFSで各レベルを左から右へ探索して値をスライスに溜める。
+レベル(0-index)が奇数だったらスライスをreverseする。
+- 気にしたこと: slices.Reverseの処理が重くならないかどうか確認した。
+slices.Reverseは線形時間で実行できる。
+入力条件より、木のノード数は高々2000で、最大ノード数を持つレベルのノード数は1000未満。
+要素数1000のスライスに対するReverseは、CPUのクロック周波数1e8psを考慮すると、1000 / 1e8 = 1e-5 = 10μs。
+これだけ短い時間で処理できるなら大丈夫。
+- n: ノード数, k: 最大ノード数を持つレベルのノード数
+    - 時間計算量: O(nk)
+    - 空間計算量: O(n)
+
+```Go
+func zigzagLevelOrder(root *TreeNode) [][]int {
+	if root == nil {
+		return [][]int{}
+	}
+	zigzagLevelNodes := [][]int{}
+	level := 0
+	currentLevelNodes := []*TreeNode{root}
+	for len(currentLevelNodes) > 0 {
+		currentLevelNodeValues := []int{}
+		nextLevelNodes := []*TreeNode{}
+		for _, node := range currentLevelNodes {
+			currentLevelNodeValues = append(currentLevelNodeValues, node.Val)
+			if node.Left != nil {
+				nextLevelNodes = append(nextLevelNodes, node.Left)
+			}
+			if node.Right != nil {
+				nextLevelNodes = append(nextLevelNodes, node.Right)
+			}
+		}
+		if level%2 == 1 {
+			slices.Reverse(currentLevelNodeValues)
+		}
+		zigzagLevelNodes = append(zigzagLevelNodes, currentLevelNodeValues)
+		level++
+		currentLevelNodes = nextLevelNodes
+	}
+	return zigzagLevelNodes
+}
+```
+
+### Step 2
+#### 2a reverse実装
+```Go
+func zigzagLevelOrder(root *TreeNode) [][]int {
+	if root == nil {
+		return [][]int{}
+	}
+	zigzagLevelNodes := [][]int{}
+	level := 0
+	currentLevelNodes := []*TreeNode{root}
+	for len(currentLevelNodes) > 0 {
+		currentLevelNodeValues := []int{}
+		nextLevelNodes := []*TreeNode{}
+		for _, node := range currentLevelNodes {
+			currentLevelNodeValues = append(currentLevelNodeValues, node.Val)
+			if node.Left != nil {
+				nextLevelNodes = append(nextLevelNodes, node.Left)
+			}
+			if node.Right != nil {
+				nextLevelNodes = append(nextLevelNodes, node.Right)
+			}
+		}
+		if level%2 == 1 {
+			// slices.Reverse(currentLevelNodeValues)
+			reverse(&currentLevelNodeValues)
+		}
+		zigzagLevelNodes = append(zigzagLevelNodes, currentLevelNodeValues)
+		level++
+		currentLevelNodes = nextLevelNodes
+	}
+	return zigzagLevelNodes
+}
+
+func reverse(s *[]int) {
+	i := 0
+	j := len(*s) - 1
+	for i < j {
+		(*s)[i], (*s)[j] = (*s)[j], (*s)[i]
+		i++
+		j--
+	}
+}
+```
+
+#### 2b reverse呼ばない
+- 参考: https://github.com/seal-azarashi/leetcode/pull/26/files#diff-39117f806e1c76b27224df2000d2200427b950646feb7d63be476d764d29dfa8R14
+- 関数呼び出しのオーバーヘッドがなくなる
```

Comment:
手元で実験してみたところ、4nsという数字が出てきました
実験方法は、slices.Reverseの実行開始から終了までの時間と、slices.Reverseの中身の開始から終了までの時間をそれぞれ計測し、差分を出しました。後者の場合の方が時間がかかる場合もあり、かなり分散が大きかったので、実質差はないだろうというのが結論になりました

```
start := time.Now()
slices.Reverse(currentLevelNodeValues)
functionCallTime += time.Since(start)
```

```
start := time.Now()
for i, j := 0, len(currentLevelNodeValues)-1; i < j; i, j = i+1, j-1 {
	currentLevelNodeValues[i], currentLevelNodeValues[j] = currentLevelNodeValues[j], currentLevelNodeValues[i]
}
functionCallTime += time.Since(start)
```

Reply:
遠い記憶を掘り出してt検定をしてみたらp値0.95 > 0.05で両者に有意差がないという帰無仮説が受諾されました

Reply:
そうですね。
適当に検索して出てきたこれも 2 ns くらいといっていますね。
https://billglover.me/2018/09/17/how-expensive-is-a-go-function-call/